src/SourceCode.Clay.Algorithms/TreeNode.cs

```diff
@@ -18,57 +18,66 @@ namespace SourceCode.Clay.Algorithms
     {
+        private readonly IEqualityComparer<T> _equalityComparer;
+
         public IReadOnlyList<int> Hierarchy { get; }
+
         public T Node { get; }
 
         public string HierarchyPath
         {
             get
             {
                 if (Hierarchy == null) return null;
+
                 var sb = new StringBuilder(1 + Hierarchy.Count * 2);
                 sb.Append("/");
                 for (var i = 0; i < Hierarchy.Count; i++)
                     sb.AppendFormat(CultureInfo.InvariantCulture, "{0}/", Hierarchy[i]);
+
                 return sb.ToString();
             }
         }
 
-        public TreeNode(T node, IReadOnlyList<int> hierarchy)
+        public TreeNode(T node, IEqualityComparer<T> equalityComparer, IReadOnlyList<int> hierarchy)
         {
             Hierarchy = hierarchy ?? throw new ArgumentNullException(nameof(hierarchy));
             Node = node;
+            _equalityComparer = equalityComparer ?? EqualityComparer<T>.Default;
         }
 
-        public TreeNode(T node, params int[] hierarchy)
-            : this(node, (IReadOnlyList<int>)hierarchy)
-        {
-        }
+        public TreeNode(T node, IEqualityComparer<T> equalityComparer, params int[] hierarchy)
+            : this(node, equalityComparer, (IReadOnlyList<int>)hierarchy)
+        { }
 
-        public static bool operator ==(TreeNode<T> node1, TreeNode<T> node2) => node1.Equals(node2);
+        public TreeNode(T node, IReadOnlyList<int> hierarchy)
+            : this(node, null, hierarchy)
+        { }
 
-        public static bool operator !=(TreeNode<T> node1, TreeNode<T> node2) => !(node1 == node2);
+        public TreeNode(T node, params int[] hierarchy)
+            : this(node, null, (IReadOnlyList<int>)hierarchy)
+        { }
 
-        public TreeNode<T> Clone()
-        {
-            if (Hierarchy is null) return default;
-            return new TreeNode<T>(Node, Hierarchy.ToArray());
-        }
+        public TreeNode<T> Clone() 
+            => Hierarchy is null ? 
+            default : 
+            new TreeNode<T>(Node, _equalityComparer, Hierarchy.ToArray()); // Clone via ToArray
 
         public override string ToString() => $"{HierarchyPath} {Node}";
 
         public override bool Equals(object obj)
-            => obj is TreeNode<T> o && Equals(o);
+            => obj is TreeNode<T> other
+            && Equals(other);
 
         public bool Equals(TreeNode<T> other)
         {
             if (Hierarchy is null) return other.Hierarchy is null;
             if (other.Hierarchy is null) return false;
 
-            if (!EqualityComparer<T>.Default.Equals(Node, other.Node)) return false;
             if (Hierarchy.Count != other.Hierarchy.Count) return false;
+            if (!_equalityComparer.Equals(Node, other.Node)) return false;
 
-            for (var i = Hierarchy.Count - 1; i >= 0; i--)
+            for (var i = 0; i < Hierarchy.Count; i++)
```

Comment:
Not sure why it was reversed. Not idiomatic but you may have had a reason

Reply:
Hierarchies have a greater probability of being different starting from the end. Please revert.